core/src/display_object/edit_text.rs

```diff
@@ -2312,3 +2328,210 @@ impl TextSelection {
+
+        Self {
+            value: Some(string.into()),
+            allowed,
+            disallowed,
+        }
+    }
+
+    fn tokenize_restrict(string: &WStr) -> VecDeque<EditTextRestrictToken> {
+        let mut characters: VecDeque<char> = string
+            .chars()
+            .map(|c| c.unwrap_or(char::REPLACEMENT_CHARACTER))
+            .collect::<VecDeque<char>>();
+        let mut tokens: VecDeque<EditTextRestrictToken> = VecDeque::with_capacity(characters.len());
+
+        while !characters.is_empty() {
+            match characters.pop_front().unwrap() {
+                // Handle escapes: \\, \-, \^.
+                // In fact, other escapes also work, so that \a is equivalent to a, not to \\a.
+                '\\' => {
+                    if let Some(escaped) = characters.pop_front() {
+                        tokens.push_back(EditTextRestrictToken::Char(escaped));
+                    } else {
+                        // Ignore truncated escapes (when the string ends with \).
+                    }
+                }
+                '^' => {
+                    tokens.push_back(EditTextRestrictToken::Caret);
+                }
+                '-' => {
+                    tokens.push_back(EditTextRestrictToken::Range);
+                }
+                c => {
+                    tokens.push_back(EditTextRestrictToken::Char(c));
+                }
+            }
+        }
+
+        tokens
+    }
+
+    fn parse_restrict(
+        tokens: &mut VecDeque<EditTextRestrictToken>,
+        allowed: &mut Vec<(char, char)>,
+        disallowed: &mut Vec<(char, char)>,
+    ) {
+        let mut current_intervals: Vec<(char, char)> = vec![];
+        let mut last_char: Option<char> = None;
+        let mut now_allowing = true;
+        while !tokens.is_empty() {
+            last_char = match tokens.pop_front().unwrap() {
+                EditTextRestrictToken::Char(c) => {
+                    current_intervals.push((c, c));
+                    Some(c)
+                }
+                EditTextRestrictToken::Caret => {
+                    if now_allowing {
+                        if current_intervals.is_empty() && allowed.is_empty() {
+                            // If restrict starts with ^, we are assuming that
+                            // all characters are allowed and disallowing from that.
+                            allowed.append(&mut vec![Self::INTERVAL_ALL]);
+                        } else {
+                            allowed.append(&mut current_intervals);
+                        }
+                    } else {
+                        disallowed.append(&mut current_intervals);
+                    }
+
+                    // Caret according to the documentation indicates
+                    // that we are now disallowing characters.
+                    // In reality it just switches allowing/disallowing.
+                    now_allowing = !now_allowing;
```

Comment:
lol, nice misdocumentation, and great that you discovered it
(just babbling, no action necessary)

Reply:
TBH it looks like an implementation detail leak. Someone at Macromedia did not think more than one caret may be included in restrict :smiley:

Reply:
Yep. And we have seen _a lot_ of these kinds of leaks.

Reply:
So, thanks, Hyrum's Law...
https://www.hyrumslaw.com/